File: app/src/main/java/com/gnoemes/shikimori/utils/network/HtmlResponseBodyConverter.java
package com.gnoemes.shikimori.utils.network;

import android.support.annotation.NonNull;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Converter;

public class HtmlResponseBodyConverter implements Converter<ResponseBody, Document> {

    private String baseUrl;

    HtmlResponseBodyConverter(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Override
    public Document convert(@NonNull ResponseBody value) throws IOException {
        try {
            return Jsoup.parse(value.byteStream(), "UTF-8", baseUrl);
        } finally {
            value.close();
        }
    }
}

File: app/src/main/java/com/gnoemes/shikimori/utils/network/HtmlResponseBodyConverter.kt
package com.gnoemes.shikimori.utils.network

import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import retrofit2.Converter

class HtmlResponseBodyConverter(
        private val baseUrl : String
) : Converter<ResponseBody, Document> {

    override fun convert(value: ResponseBody): Document {
        return value.use { value ->
            Jsoup.parse(value.byteStream(), "UTF-8", baseUrl)
        }
    }
}